### ISSUE TITLE
Tabs

### ISSUE THREAD

**Issue by GoogleCodeExporter** [mintty/mintty#8]
```
I would like to see an option to "Allow only one instance." This option
would subsequently suck all new terminal sessions into a new tab in the
same window.  The user could then switch tabs with a customizable key
combination, that would default to something reasonable like Ctrl+Tab.
```

Original issue reported on code.google.com by `forrest....@gmail.com` on 30 Dec 2008 at 4:45

**Comment by laktak** [comment]
@mintty since the original *wont-fix* comment is from 2010 -- Is your opinion still the same or will we see tabs in the future?

**Comment by mintty** [comment]
Well, “your” isn’t exactly the same person as it used to be... but I share Andy’s opinion about where and how Tabs should be implemented, see https://github.com/mintty/mintty/issues/8#issuecomment-108886260 and https://github.com/mintty/mintty/issues/8#issuecomment-108886297.

An implementation intertwined with the current code is not acceptable because it might comprise the stability of the terminal.
If at all, there could be some kind of add-on that runs multiple instances of the terminal code in widely decoupled processes.

**Comment by laktak** [comment]
In that case, would officially supporting [ConEmu](http://conemu.github.io/) be an option? 

I'm using it with mintty at the moment and it appears to work quite well.

**Comment by mintty** [comment]
I’ve had another look at ConEmu on this occasion and I don’t currently have the nerve to investigate its configuration which isn’t quite - too intuitive.
If you’d like to contribute a mintty-specific pre-configuration for ConEmu, that would indeed be appreciated. Something like a ConEmu configuration file and a desktop shortcut to start an instance of ConEmu with mintty, any maybe another demo shortcut showing how to start two instances...

**Comment by laktak** [comment]
It's actually quite easy:

- Download [ConEmu Stable, Portable](http://www.fosshub.com/ConEmu.html)
- Start ConEmu and set the *startup task* to `{Bash::Git bash}` on the first dialog that appears

You'll get a window with mintty/bash running and can create more tabs with `<win> + w` or by clicking on `+`.

`TERM` is set to `cygwin` ... I think I had some issues with that as I set mine to `xterm`.

**Comment by mintty** [comment]
If TERM=cygwin, you have been starting a Cygwin Console in ConEmu, not a Cygwin Terminal (i.e. mintty).

**Comment by roeme** [comment]
IMO, considering Tabs only makes sense after #56 is resolved. With the current state of affairs using a terminal muxer like `tmux` or `screen` aligns better with mintty's goals; that is, being a lightweight terminal emulator (Even more so since personally I can use most of the few Windows CLI tools I need within mintty/tmux).

Also, with today's ever increasing screen estate; I'm tempted to claim that the majority uf users prefer more terminal windows per screen, instead of Tabs.

**Comment by ryanerwin** [comment]
Wow... I'm really amazed at how far `mintty` has come recently... True color. Sixel. WSL.

I've been using ConEmu and Tmux for a long time, but:

1) Tmux doesn't have Sixel support, and doesn't plan on adding it
2) ConEmu has developed their own Cygwin wrapper that I think they're most focused on as opposed to "child window" support... And ConEmu support of `mintty` for resizing windows, pressing Alt+Enter don't work properly.

**Comment by mintty** [comment]
Also neither ConEmu nor tmux are themselves terminals.

**Comment by mintty** [comment]
I had some intuition how to conciliate the issue and implemented "Virtual Tabs". A proof of concept demo is available for download. A menu displays a switching list of all mintty windows.
For the demo, it's attached to the context menu and also reachable with Ctrl+Shift+I.
Other keys might be desirable, and probably it should not be in the default context menu, but maybe a separate one, to open with some modifier+mouse.

**Comment by mkq** [comment]
Thanks.

"Available for download" is a bit misleading IMHO. You mean one should download a snapshot of master and build oneself? Or am I missing some snapshot build link?

Anyway, for me, the main benefit of tabs would be to notice when another terminal's window title (which should obvioulsy be used as tab title) changes. I.e. I'd immediatley notice when another shell is back at the prompt after a long running command. So "tabs" hidden in a popup menu unfortunately don't cut it for me.

(Since I usually have only two or three minttys open, the focus switching aspect of tabs is already handled decently with Ctrl+Tab. Or most often, I prefer some Autohotkey shortcuts.)

Could you maybe add an option that makes the mintty window title show all other minttys' titles too (with fixed positions in launch order)?
`+-+---------------------------------------------------------------+----+`
`|·|·launched·1st··|·launched·2nd··|·focussed·3rd··|·last·launched·|―[]X|`
`+-+-------------------------------^^^^^^^^^^^^^^^^^---------------+----+`
`|······································································|`
`|·/tmp·>·ls_···························································|`
`|······································································|`
`|······································································|`
`+----------------------------------------------------------------------+`
(The dots should be spaces, but multiple consecutive spaces where not displayed correctly.)

The focussed windows's title should somehow be highlighted. Is underlining possible in a title bar?
I think I'd prefer when non-focussed windows retained their original titles. But if it's all or nothing, changing all windows would be ok, too.

**Comment by mintty** [comment]
Here's the challenge: It is not possible to set any style in the title bar. It is possible to detect mouse clicks, but it is hardly possible to correlate them to the proper field because there is no control about the font of the window title (and accordingly the spacing).
Maybe, speculating, it could be possible to mirror the title layout by fiddling out the font and size being used and render a copy of the title in memory to detect the effective proportions and handle the border positions. Maybe it would be even easier to self-draw the title, which is the ultimate alternative for applications that want to run their own thing here.

**Comment by mintty** [comment]
Released 2.8.0.

**Comment by mintty** [comment]
> immediatley notice when another shell is back at the prompt

How about making the foreground window flash (e.g. when invoking a bell)?

**Comment by mkq** [comment]
> How about making the foreground window flash (e.g. when invoking a bell)?

Nice, I had forgotten about that. I made an [Autohotkey DllCall "FlashWindow"](https://autohotkey.com/board/topic/42474-is-it-possible-to-make-a-window-flash/) based hack which I can now retire. Thanks!

**Comment by mintty** [comment]
Actually, the foreground flash is not yet implemented, I was just asking. You could of course send either a BEL or a title setting sequence to the tty of the foreground terminal, but then you'd need to find out which...

**Comment by mintty** [comment]
You could also send a BEL to the own session after running a command, and configure `BellPopup=yes`,
but I guess that's not what you want (immediate popup and having to send something explicitly).

**Comment by mkq** [comment]
> You could of course send either a BEL or a title setting sequence to the tty of the foreground terminal, but then you'd need to find out which...

Sorry, I don't understand why I would need to find out which terminal. But anyway...

> You could also send a BEL to the own session after running a command

... that's what I did if I understand you correctly. I.e. I do `print -Pn '\a'` in zsh's `precmd` function (`echo -ne '\a'` in `$PS1` in bash should have the same effect).
With mintty option "Highlight in taskbar" on (`BellTaskbar=yes`), the taskbar button gets highlighted only when not focussed, which is what I wanted. Thanks!

**Comment by mintty** [comment]
If that's sufficient for you, the better, and there's no inter-session communication needed.
I had assumed you would wish the current foreground session to become somehow indicative, and that would need to introduce an inter-session notification mechanism.

**Comment by mkq** [comment]
Yes and no :-) Sorry I wasn't clear on that. I meant and still think, to make mintty feel like a tabbed terminal, some inter-session notification would be required. And that notification should be apparent, not just in a context menu. Thanks for the hint to redirect a title setting sequence to another terminal device. I hadn't thought about that. Maybe I'll experiment with that later.

OTOH, starting with your "Here's the challenge..." answer, I was talking about the "workaround" to make separate windows more background-friendly.

**Comment by mintty** [comment]
OK, here's the deal: If someone provides a graphical tab bar implementation (e.g. using winsearch.c as a pattern), I will provide the functional background including any inter-session communication as needed.

**Comment by mkq** [comment]
Cool. Though that someone is not me, as I suck at C programming.

**Comment by mintty** [comment]
I'm closing this issue now that Virtual Tabs are available.
For remaining enhancements, see the follow-up issue #699.
For implementation constraints, see also https://github.com/mintty/mintty/issues/600#issuecomment-344165427.

**Comment by mintty** [comment]
Released 3.0.3 with title-bar session switcher (Ctrl+right-click) which is the prototype of an optical tab bar.

**Comment by dxhisboy** [comment]
@mintty I'm trying to implement a tab bar based by adding a component like the search bar on the top.
![image](https://user-images.githubusercontent.com/3888709/68990537-40479500-088f-11ea-84e5-4485eebdf13d.png)
Now I have some questions:
1. Any suggestions on selecting the colors? Also, is there a suggested way to get the background and foreground color of the terminal?
2. Should the tabs in the bar have a fixed width or dynamic width when a tab has a very long title? If a fixed width is wanted, which part of the title should I display?

**Comment by mintty** [comment]
1. Within mintty, a number of configured colours are available: cfg.fg_colour, cfg.bg_colour, cfg.cursor_colour, cfg.search_fg_colour, cfg.search_bg_colour, cfg.search_current_colour, cfg.sel_fg_colour, cfg.sel_bg_colour. New entries could be defined, perhaps with a fallback to some of the above if not configured (-1).Within mintty, a number of configured colours are available: cfg.fg_colour, cfg.bg_colour, cfg.cursor_colour, cfg.search_fg_colour, cfg.search_bg_colour, cfg.search_current_colour, cfg.sel_fg_colour, cfg.sel_bg_colour. New entries could be defined, perhaps with a fallback to some of the above if not configured (-1).
2. Other applications (e.g. browsers) vary width with the number of tabs but not with the length of contents. They usually display the initial part; considering that a terminal title will often display the working directory, it may be a useful option to display the end instead.
3. The tab set and contents should be updated if terminal windows are created/closed/change title. A framework for this interaction is already available, as used by the "session switcher", but it may not yet be complete for active updates; I would contribute missing functionality.
4. Typically you would expect a set of tabs in one bar. As a general alternative to an inline tab bar, a tab bar could be displayed above the window. I had once thought about a separate process that adjusts its window position accordingly and might be implemented as a special mode of mintty. You are aware that something similar is already available, just not always displayed but to be shown with Ctrl+click on the title bar?

**Comment by dxhisboy** [comment]
Yes, I found the code and I'm currently forcing geomsync to let the window notify each other.

**Comment by dxhisboy** [comment]
I currently use first letter + "*" + suffix as long as possible for too long titles and the background and foreground color is set to `ansi_colours[10]` and `ansi_colours[0]`. Will this be okay?
![image](https://user-images.githubusercontent.com/3888709/69028402-d8b75400-0a0c-11ea-8a09-ec0944b34904.png)
Then I will process the dynamic tab length, and necessary interactions.

**Comment by mintty** [comment]
In this screenshot, the tab looks almost like normal terminal output from an application.
This may be confusing. Style should be clearly distinctive. More decent colours, perhaps, maybe a thin separator bar, and a smaller font (it looks even larger than the text above, should be the other way).

**Comment by mintty** [comment]
I think some system colours should be used for the tab display, e.g. the title bar colour for the active tab and a dimmed version for the others, or the title colour for others and some menu or info colour for the active tab background. Colour values can be retrieved with GetSysColor.

**Comment by dxhisboy** [comment]
I think you are right, I found this looks like a tmux tabbar.

**Comment by dxhisboy** [comment]
![image](https://user-images.githubusercontent.com/3888709/69039227-78351080-0a26-11ea-9597-f408ef405dd4.png)
I think I can follow this color scheme, but they do not look alike even they have same RGB value...

**Comment by BrianInglis** [comment]
> I currently use first letter + "*" + suffix as long as possible for too long titles
Would be better to use an ellipsis character (below) instead of `*`: 
`…  U+02026 HORIZONTAL ELLIPSIS`
`⋯  U+022EF MIDLINE HORIZONTAL ELLIPSIS`

**Comment by dxhisboy** [comment]
@BrianInglis `… U+02026 HORIZONTAL ELLIPSIS` works.
![image](https://user-images.githubusercontent.com/3888709/69110223-ffc96080-0ab4-11ea-97b3-4b85104a9fac.png)
But a potential problem is the `⋯ U+022EF MIDLINE HORIZONTAL ELLIPSIS` is missing in my font:
![image](https://user-images.githubusercontent.com/3888709/69110262-1f608900-0ab5-11ea-89a3-8cc43daef753.png)
Then how could we ensure the user is using a supported font become a problem.

**Comment by BrianInglis** [comment]
Most fonts should support the standard publishing marks: 
`․  U+02024 ONE DOT LEADER`
`‥ U+02025 TWO DOT LEADER`
`… U+02026 HORIZONTAL ELLIPSIS` aka 3 dot leader
`⋮U+022EE VERTICAL ELLIPSIS`
but perhaps not some of the others.

**Comment by dxhisboy** [comment]
It is now almost working. Just one question:
Use windows' tab control as the graphics component or paint it by myself?
I have both code working.

**Comment by mintty** [comment]
Don't see what you mean by "graphics" versus "paint". Maybe I should try it out...

**Comment by dxhisboy** [comment]
@mintty I've pushed to https://github.com/dxhisboy/mintty/tree/dev
git diff output is here: [diff.txt](https://github.com/mintty/mintty/files/3867674/diff.txt)
Add `ShowTabBar=yes` to minttyrc to enable it.
Code still needs heavy polishing.

**Comment by mintty** [comment]
This seems to work well at first glance, amazing how you got the session framework integrated without specific support...
I have some initial formal review, mainly concerning code layout, which I'll add as a commit comment.
Tabbar layout can also do with some tweaking, I suppose. The Tab separation is currently not very visible, the font could be smaller, I think.

**Comment by dxhisboy** [comment]
What does tab separation mean? The "separation between the tabbar and terminal" or "the separation between adjacent tabs"?

**Comment by Biswa96** [comment]
How the tab feature work here? I have to create a new window to see a new tab. Should it not be just one window and multiple tab?

Some points about patch file: 

* file permissions are changed from `644` to `755`.
* patch contains tab and space both.
* add a space b/w `)` and `{`. e.g. `if (x ==0) {` and `void func(int x) {`.
* device contexts are not released after used.

**Comment by dxhisboy** [comment]
> How the tab feature work here? I have to create a new window to see a new tab. Should it not be just one window and multiple tab?
Follows the consideration of "Virtual Tabs", this is a "Virtual Tabbar".

**Comment by dxhisboy** [comment]
A newer version is available: https://github.com/dxhisboy/mintty/tree/dev and [diff.txt](https://github.com/mintty/mintty/files/3869296/diff.txt).

**Comment by dxhisboy** [comment]
I'm trying a smaller font (`0.8*cell_height`), but it looks strange.
![image](https://user-images.githubusercontent.com/3888709/69241589-10272b80-0bda-11ea-9fc7-7ec412bedb72.png)

**Comment by dxhisboy** [comment]
Are these two border styles better?
![image](https://user-images.githubusercontent.com/3888709/69243429-03a4d200-0bde-11ea-827b-e35dd88a6d5b.png)
![image](https://user-images.githubusercontent.com/3888709/69243458-15867500-0bde-11ea-8648-574affac49d9.png)

**Comment by mintty** [comment]
Border style looks nicer, thanks.
Tabbar text is placed higher for active tab, for me that looks distracting.
For the active tab, I would try to use the same colour is in the title bar, because these are related; don't know whether it'll look good.

**Comment by Biswa96** [comment]
> Actually one window and multiple tab will be very complex.

First comment of this thread mentions this:

> This option would subsequently suck all new terminal sessions into a new tab in the same window.

**Comment by mintty** [comment]
> Also an experimental feature is if we click the tab for window B in window A, then window B will be moved to the position of window A.

That is actually SessionGeomSync=1 or higher already and shouldn't be implied without setting SessionGeomSync. We can later talk about unifying options.

**Comment by BrianInglis** [comment]
> It is now almost working. Just one question:
> Use windows' tab control as the graphics component or paint it by myself?
> I have both code working.

Always let Windows do it so they are responsible for matching style changes.

**Comment by BrianInglis** [comment]
> I'm trying a smaller font (`0.8*cell_height`), but it looks strange.

Might also want to skip display of characters after "`…`" before "`/`"; if no "`/`" displayed but "`-`", then after "`…`" before "`-`"; if neither "`/`" nor "`-`" displayed but "`_`", then after "`…`" before "`_`"; unless the number of characters skipped would equal or exceed the number of characters remaining

**Comment by k-takata** [comment]
How about creating a new WIP PR instead of keep discussing in an already closed issue?

**Comment by mintty** [comment]
Could have been a separate issue, yes. Up to @dxhisboy whether to create a new issue, doesn't need to be a PR while being developed and reviewed.
Unfortunately, github does not seem to have a feature to split an issue / move comments (I sent them a request).

**Comment by BrianInglis** [comment]
Would be nice to support git operations on issues - branch, fork, cherry-pick, stash, merge, migrate
